Mathlib/Algebra/Category/ModuleCat/ChangeOfRings.lean

```diff
@@ -852,6 +877,107 @@ instance preservesColimitRestrictScalars {R S : Type*} [Ring R] [Ring S]
   apply isColimitOfPreserves (forget₂ (ModuleCat.{v} S) AddCommGrp.{v})
   exact HasColimit.isColimitColimitCocone F
 
+variable (R) in
+/-- The extension of scalars by the identity of a ring is isomorphic to the
+identity functor. -/
+noncomputable def extendScalarsId : extendScalars (RingHom.id R) ≅ 𝟭 _ :=
+  ((conjugateIsoEquiv (extendRestrictScalarsAdj (RingHom.id R)) Adjunction.id).symm
+    (restrictScalarsId R)).symm
+
+lemma extendScalarsId_inv_app_apply (M : ModuleCat R) (m : M):
+    (extendScalarsId R).inv.app M m = (1 : R) ⊗ₜ m := rfl
+
+lemma homEquiv_extendScalarsId (M : ModuleCat R) :
+    (extendRestrictScalarsAdj (RingHom.id R)).homEquiv _ _ ((extendScalarsId R).hom.app M) =
+      (restrictScalarsId R).inv.app M := by
+  ext m
+  rw [extendRestrictScalarsAdj_homEquiv_apply, ← extendScalarsId_inv_app_apply]
+  erw [← comp_apply]
+  simp
+
+lemma extendScalarsId_hom_app_one_tmul (M : ModuleCat R) (m : M) :
+    (extendScalarsId R).hom.app M ((1 : R) ⊗ₜ m) = m := by
+  rw [← extendRestrictScalarsAdj_homEquiv_apply,
+    homEquiv_extendScalarsId]
+  dsimp
+
+section
+
+variable {R₁ R₂ R₃ R₄ : Type u₁} [CommRing R₁] [CommRing R₂] [CommRing R₃] [CommRing R₄]
+  (f₁₂ : R₁ →+* R₂) (f₂₃ : R₂ →+* R₃) (f₃₄ : R₃ →+* R₄)
+
+/-- The extension of scalars by a composition of commutative ring morphisms
+identify to the composition of the extension of scalars functors. -/
+noncomputable def extendScalarsComp :
+    extendScalars (f₂₃.comp f₁₂) ≅ extendScalars f₁₂ ⋙ extendScalars f₂₃ :=
+  (conjugateIsoEquiv
+    ((extendRestrictScalarsAdj f₁₂).comp (extendRestrictScalarsAdj f₂₃))
+    (extendRestrictScalarsAdj (f₂₃.comp f₁₂))).symm (restrictScalarsComp f₁₂ f₂₃).symm
+
+lemma homEquiv_extendScalarsComp (M : ModuleCat R₁) :
+    (extendRestrictScalarsAdj (f₂₃.comp f₁₂)).homEquiv _ _
+      ((extendScalarsComp f₁₂ f₂₃).hom.app M) =
+      (extendRestrictScalarsAdj f₁₂).unit.app M ≫
+        (restrictScalars f₁₂).map ((extendRestrictScalarsAdj f₂₃).unit.app _) ≫
+        (restrictScalarsComp f₁₂ f₂₃).inv.app _ := by
+  dsimp [extendScalarsComp, conjugateIsoEquiv, conjugateEquiv]
+  simp only [Category.assoc, Category.id_comp, Category.comp_id,
+    Adjunction.comp_unit_app, Adjunction.homEquiv_unit,
+    Functor.map_comp, Adjunction.unit_naturality_assoc,
+    Adjunction.right_triangle_components]
+  rfl
+
+lemma extendScalarsComp_hom_app_one_tmul (M : ModuleCat R₁) (m : M) :
+    (extendScalarsComp f₁₂ f₂₃).hom.app M ((1 : R₃) ⊗ₜ m) =
+      (1 : R₃) ⊗ₜ[R₂,f₂₃] ((1 : R₂) ⊗ₜ[R₁,f₁₂] m) := by
+  rw [← extendRestrictScalarsAdj_homEquiv_apply, homEquiv_extendScalarsComp]
+  rfl
+
+@[reassoc]
+lemma extendScalars_assoc :
+    (extendScalarsComp (f₂₃.comp f₁₂) f₃₄).hom ≫ whiskerRight (extendScalarsComp f₁₂ f₂₃).hom _ =
+      (extendScalarsComp f₁₂ (f₃₄.comp f₂₃)).hom ≫ whiskerLeft _ (extendScalarsComp f₂₃ f₃₄).hom ≫
+        (Functor.associator _ _ _).inv := by
+  ext M m
+  dsimp
+  have := extendScalarsComp_hom_app_one_tmul (f₂₃.comp f₁₂) f₃₄ M m
```

Comment:
Is this to make the code faster, or is this `have` statement useful?

Reply:
(note also that if it is for efficiency purposes, it would be unecessary once there is a `dsimp%` elaborator as I mentioned in the main comment)

Reply:
It does not work without a `have` statement. I have cleaned up a bit, there is still one `erw` here.